### ISSUE TITLE
Support for JS Functions in EvaluationEnvironment bindings.

### ISSUE THREAD

**Issue by jaugat** [mikhail-barg/jsonata.net.native#15]
I'm looking for a solution to passing in a javascript function bound to a variable via the bindings instead of using a static C# function since the maps will be used across multiple products/teams utilizing JSONata for data transformation and not all of these products are using .Net. For example, when using the JSONata exerciser at try.jsonata.org, I am able to pass in a function in the bindings section similar to this:

```
{
  "findFirst": (...args) => { return args.find(a => a) }
}
```

When passing in the following source: 

```
{
  "key1": "",
  "key2": "some value",
  "key3": "another value"
}
```

and the following map:

```
{
  "result": $findFirst(key1, key2, key3)
}
```

it returns the following result:

```
{
  "result": "some value"
}
```

However, when trying to parse the bindings into a JObject as shown below on line 2 for the EvaluationEnvironment object, I get a JsonParseException.

```
JToken source = JToken.Parse("{\"key1\": \"\", \"key2\": \"some value\", \"key3\": \"another value\"}");
JObject bindings = JToken.Parse("{\"findFirst\": (...args) => { return args.find(a => a) }}") as JObject;
EvaluationEnvironment evaluationEnvironment = new EvaluationEnvironment(bindings);
JsonataQuery evaluator = new JsonataQuery("{\"result\": $findFirst(key1, key2, key3)}");
Console.WriteLine(evaluator.Eval(source, evaluationEnvironment).ToFlatString());
```

Are JS function in bindings supported or is there another way to accomplish this aside from binding static C# functions?

**Comment by mikhail-barg** [comment]
Hi @jaugat . As for now, this package is a port of original [jsonata-js](https://github.com/jsonata-js/jsonata) to .Net, and is generally aimed to C# users/developers. So no support for JS functions out of the box is provided.

If you are looking for using JSONata language in different programming environments, you may want to use native implementation for each language. Here's a list of known existing implementations: https://github.com/jsonata-js/jsonata/blob/master/docs/overview.md#implementations-of-jsonata

If you are looking to writing some C# service/app for users that would want to use JS functions in JSONata queries, e.g. implementing try.jsonata.org as-is but on .net, I see the following options:

* you may want to use my [jsonata.net.js](https://github.com/mikhail-barg/jsonata.net.js) package which is just a wrapper over original jsonata-js code run in .Net via [Jint](https://github.com/sebastienros/jint) JS-C# interop library. This is completely unrelated to the current package, and it runs [much slower](https://github.com/mikhail-barg/jsonata.net.native#performance) than `jsonata.net.native`, but still is a working solution, though a bit outdated.

* another solution would be to use `jsonata.net.native`, but wrap just custom js functions into C# delegate, and then pass to `ExecutionEnvironment.BindFunction()`. 

If you choose latter option, you would find that there's currently no `BindFunction()` override accepting a delegate. So I'll hjave a look on implementing this.

**Comment by mikhail-barg** [comment]
Okay, i've just published a new release, v2.4.0, with support for passing delegates (and therefore lambdas) to `EvaluationEnvironment.BindFunction()`. 

Hope this would help in your case. 

Still not sure about "vararg" passing, as in your example. This may still be not possible, at least until "function signature" feature is supported.

**Comment by jaugat** [comment]
Thanks for the quick response to this! I'll start diving into the new features.
